primitives/src/circuit/rescue/native.rs

```diff
@@ -97,95 +106,7 @@ impl<F: PrimeField> Gate<F> for Power5NonLinearGate<F> {
-    /// * `returns` - state variable corresponding to the cipher text
-    fn prp(
-        &mut self,
-        key_var: &RescueStateVar,
-        input_var: &RescueStateVar,
-    ) -> Result<RescueStateVar, CircuitError>;
-
-    /// Sponge-based hashes from Rescue permutations
-    /// * `data_vars` - sponge input variables, `data_vars.len()` should be a
-    ///   positive integer that is a multiple of the sponge rate (i.e. 3)
-    /// * `num_output` - number of output variables
-    /// * `returns` - a vector of variables that refers to the sponge hash
-    ///   output
-    fn rescue_sponge_no_padding(
-        &mut self,
-        data_vars: &[Variable],
-        num_output: usize,
-    ) -> Result<Vec<Variable>, CircuitError>;
-
-    /// Sponge-based hashes from Rescue permutations
-    /// * `data_vars` - sponge input variables,
-    /// * `num_output` - number of output variables
-    /// * `returns` - a vector of variables that refers to the sponge hash
-    ///   output
-    fn rescue_sponge_with_padding(
-        &mut self,
-        data_vars: &[Variable],
-        num_output: usize,
-    ) -> Result<Vec<Variable>, CircuitError>;
-
-    /// Full-State-Keyed-Sponge with a single output
-    /// * `key` - key variable
-    /// * `input` - input variables,
-    /// * `returns` a variable that refers to the output
-    fn rescue_full_state_keyed_sponge_no_padding(
-        &mut self,
-        key: Variable,
-        data_vars: &[Variable],
-    ) -> Result<Variable, CircuitError>;
-
-    /// Return the round keys variables for the Rescue block cipher
-    /// * `mds_states` - Rescue MDS matrix
-    /// * `key_var` - state variable representing the cipher key
-    /// * `returns` - state variables corresponding to the scheduled keys
-    fn key_schedule(
-        &mut self,
-        mds_states: &RescueMatrix<F>,
-        key_var: &RescueStateVar,
-        prp_instance: &PRP<F>,
-    ) -> Result<Vec<RescueStateVar>, CircuitError>;
-
-    /// Create a variable representing a rescue state
-    /// * `state` - Rescue state
-    /// * `returns` - state variables corresponding to the state
-    fn create_rescue_state_variable(
-        &mut self,
-        state: &RescueVector<F>,
-    ) -> Result<RescueStateVar, CircuitError>;
-
-    /// Return the variable corresponding to the output of the of the Rescue
-    /// PRP where the rounds keys have already been computed "dynamically"
-    /// * `input_var` - variable corresponding to the plain text
-    /// * `mds_states` - Rescue MDS matrix
-    /// * `key_vars` - variables corresponding to the scheduled keys
-    /// * `returns` -
-    fn prp_with_round_keys(
-        &mut self,
-        input_var: &RescueStateVar,
-        mds: &RescueMatrix<F>,
-        keys_vars: &[RescueStateVar],
-    ) -> Result<RescueStateVar, CircuitError>;
-}
-
-impl<F> RescueGadget<F> for PlonkCircuit<F>
+impl<F> RescueGadget<RescueStateVar, F, F> for PlonkCircuit<F>
```

Comment:
Would it be helpful to have a type-alias or wrapper for the native/non-native gadget's API?

E.g.,  currently, we need to write: `RescueGadget::<RescueStateVar, F, F>::rescue_sponge_no_padding`.
It'll be great if the user only needs to write `RescueNativeGadget::<F>::rescue_sponge_no_padding`.

Similarly, for non-native setting, it'll be great if the user only needs to write `RescueNonNativeGadget<F, T>::rescue_sponge_no_padding`.

Reply:
Yes, that would indeed improve the ergonomics of using the gadget. I'll have a look, thx for the suggestion!